### PR TITLE
feat(scheduler): replace prequeue watcher with volcano extender plugin

### PR DESCRIPTION
**Pull request by YiD11**
This PR replaces crater's prequeue watcher with a volcano scheduler extender plugin. Volcano now calls back to crater before admitting each vcjob, and crater decides whether the job passes the user's queue quota. This gives volcano full control over scheduling while crater retains quota enforcement.

User-facing changes:

Jobs show a new Inqueue status on the frontend when waiting for resources, and the PodGroup phase is visible in job details.
The normal/backfill schedule type selector is removed from the job submission form and CLI.
Platform admins configure the scheduler extender connection instead of the old prequeue parameters.

A drain loop handles jobs still in the legacy Prequeue state so the upgrade is seamless. The Helm chart exposes the extender endpoint when schedulerExtenderPort is set.

Job status on frontend page are shown below:

image

image

### CODE REVIEW

File: backend/pkg/reconciler/vcjob-reconciler.go [hunk 1]
```diff
@@ -281,33 +301,52 @@ func (r *VcJobReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl
...
-		status == batch.Aborted ||
-		status == batch.Terminated
+// getJobStatus splits volcano's Pending into "not yet admitted" and "admitted, waiting for
+// nodes": the vcjob phase alone cannot tell them apart, the pod group phase can. The raw pod group
+// phase is returned alongside so the UI can tell "waiting for nodes" from "bound, containers starting".
+func (r *VcJobReconciler) getJobStatus(
+	ctx context.Context, job *batch.Job,
+) (batch.JobPhase, scheduling.PodGroupPhase, error) {
+	phase := job.Status.State.Phase
+	if phase != batch.Pending {
```
Comment (AkashiSensei) [inline backend/pkg/reconciler/vcjob-reconciler.go:318]: 【核心规范】这里用 `phase != batch.Pending` 直接跳过了空 phase 的 PodGroup 查询。新建作业在同步到 Volcano 前可能仍是空 phase，已经存在且已 admitted 的 PodGroup 会被记成空状态，而不是 `Inqueue`，配额账本也会少记。同文件测试覆盖了这个场景。

File: backend/cmd/gorm-gen/models/migrate.go [hunk 2]
```diff
@@ -1573,6 +1573,57 @@ func main() {
 			},
 		},
 		modelDownloadSubmissionMigration(),
+		{
+			ID: "202609011000",
+			Migrate: func(tx *gorm.DB) error {
+				return dropTableIfPresent(tx, "prequeue_configs")
```
Comment (AkashiSensei) [inline backend/cmd/gorm-gen/models/migrate.go:1579]: 【核心规范】`202609011000` 直接删除 `prequeue_configs`，没有把已有的 `queue_quota_enabled` 和 `normal_job_waiting_tolerance_seconds` 迁到 `system_config`。升级后，线上已打开的配额会回到默认关闭，等待忍耐时间也会回到 300 秒。

File: cli/internal/api/job.go [hunk 3]
```diff
@@ -88,10 +87,10 @@ type JobInfo struct {
 	Owner                   string       `json:"owner"`
 	UserInfo                UserInfo     `json:"userInfo"`
 	JobType                 string       `json:"jobType"`
-	ScheduleType            int          `json:"scheduleType"`
 	WaitingToleranceSeconds *int64       `json:"waitingToleranceSeconds,omitempty"`
 	Queue                   string       `json:"queue"`
 	Status                  string       `json:"status"`
+	PodGroupPhase           string       `json:"podGroupPhase,omitempty"`
```
Comment (AkashiSensei) [inline cli/internal/api/job.go:93]: 【核心规范】CLI 消费的作业结构在这里新增了 `podGroupPhase`，并去掉了 `scheduleType`；创建请求和管理配置路径也一起变了。后端与 CLI 的 `APIVersion` 仍为 1，PR 里也没有最低支持版本的结论。根 CONTRIBUTING 要求这类 CLI 契约变化同步递增两端版本，并记录最低支持版本是否调整。旧 CLI 的 `scheduleType` 是 `int`，新后端不再返回该字段时会解成 `0`（旧语义为 backfill）。

File: cli/docs/COMMANDS.md [hunk 4]
```diff
@@ -439,14 +439,13 @@ CLI 发出的平台请求带 `User-Agent: crater-cli/<product-version>` 与 `X-C
...
   - `--type` (string slice): 服务端过滤作业类型，可重复或逗号分隔，最多 20 项；类型为 `jupyter | webide | custom | pytorch | tensorflow | kuberay | deepspeed | openmpi`。
-  - `--schedule` (string slice): 服务端过滤调度类型，可重复或逗号分隔，值为 `normal | backfill`，最多 20 项。
   - `--node` (string): 服务端过滤运行在指定节点上的作业。
   - `--owner` (string): 本地按用户名或作业响应中的 owner 精确筛选。
   - `--from` / `--to` (string): 本地按 `createdAt` 时间范围筛选，支持 RFC3339 或 `YYYY-MM-DD`。
   - `--interactive` (bool): 服务端只返回交互式作业（`jupyter` / `webide`）。
   - `--batch` (bool): 服务端只返回非交互式作业。
   - `--page` (int, default `1`) / `--page-size` (int, default `15`, max `200`): 请求指定服务端分页。
-  - `--sort` (string): 最多 3 个逗号分隔的服务端排序字段，字段前加 `-` 表示降序；支持 `name | jobName | owner | queue | jobType | scheduleType | status | billedPointsTotal | createdAt | startedAt | completedAt`，不允许重复字段。
+  - `--sort` (string): 最多 3 个逗号分隔的服务端排序字段，字段前加 `-` 表示降序；支持 `name | jobName | owner | queue | jobType | status | billedPointsTotal | createdAt | startedAt | completedAt`，不允许重复字段。
```
Comment (AkashiSensei) [inline cli/docs/COMMANDS.md:448]: 【核心规范】作业创建已经不再支持 schedule type，但第 575 行仍写 TensorFlow / PyTorch 不允许 `scheduleType=0`。同一文档第 746 行仍列出 `crater context prequeue`，第 791 行仍列出 `crater admin system-config ...|prequeue`。这些接口和字段已经不存在。

File: backend/internal/handler/system_config.go [hunk 5]
```diff
@@ -63,8 +60,8 @@ func (mgr *SystemConfigMgr) RegisterAdmin(g *gin.RouterGroup) {
 
 	g.GET("/gpu-analysis", mgr.GetGpuAnalysisStatus)
 	g.PUT("/gpu-analysis", mgr.SetGpuAnalysisStatus)
-	g.GET("/prequeue", mgr.GetPrequeueConfig)
-	g.PUT("/prequeue", mgr.UpdatePrequeueConfig)
+	g.GET("/scheduler-extender", mgr.GetSchedulerExtenderConfig)
```
Comment (AkashiSensei) [inline backend/internal/handler/system_config.go:63]: 【核心规范】管理员配置已经换成 `/scheduler-extender`，但 `website/content/docs/admin/more/prequeue.*.mdx` 仍在描述预排队、回填作业和激活扫描。产品文档和当前实现不一致。

File: frontend/src/routes/admin/more/-components/scheduler-extender-settings.tsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            min={1}
+            value={waitingToleranceSeconds}
+            onChange={(event) => onWaitingToleranceSecondsChange(event.target.value)}
+            disabled={isPending}
+            placeholder={t('systemConfig.schedulerExtender.waitingTolerancePlaceholder')}
+          />
+        </div>
+      </CardContent>
+      <CardFooter className="bg-muted/10 justify-end border-t px-6 py-4">
+        <Button type="button" onClick={onSubmit} disabled={isPending}>
```
Comment (AkashiSensei) [inline frontend/src/routes/admin/more/-components/scheduler-extender-settings.tsx:122]: 【核心规范】保存会立刻改全局准入、配额和超时策略，属于非幂等的管理配置变更，当前没有确认弹窗说明影响范围。同页的 Pod 带宽和模型下载限额在 mutation 前都有确认框。

File: frontend/src/routes/portal/jobs/new/jupyter-job.tsx [hunk 7]
```diff
@@ -108,7 +106,6 @@ const formSchema = z.object({
   alertEnabled: z.boolean().default(true),
   cpuPinningEnabled: z.boolean().default(false),
   forwards: forwardsSchema,
```
Comment (AkashiSensei) [inline frontend/src/routes/portal/jobs/new/jupyter-job.tsx:108]: 【核心规范】Jupyter / WebIDE / Custom 模板序列化都删除了 `scheduleType`，但 `MetadataForm*` 仍使用 `CurrentJobTemplateVersion = '20260707'`，现有迁移链也只处理 `nodeSelector`。带有旧字段、版本号仍是当前版本的模板不会经过迁移，版本契约和实际 schema 不一致。

File: backend/pkg/extender/server.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+	"github.com/raids-lab/crater/pkg/utils"
+)
+
+const (
+	readHeaderTimeout = 5 * time.Second
+	shutdownTimeout   = 5 * time.Second
+	maxRequestBytes   = 10 << 20
+	// sessionStaleAfter only covers a lost round-close notification: volcano sends that one and
+	// forgets it, never retrying. Normal invalidation is driven by the notification itself.
+	sessionStaleAfter = 2 * time.Second
```
Comment (AkashiSensei) [inline backend/pkg/extender/server.go:26]: 【优化建议】`sessionStaleAfter = 2s` 会对仍在进行的 scheduling session 重建快照，并不只在 `onSessionClose` 丢失时失效。单轮调度超过 2 秒时，同一轮里不同作业会按不同的配额底座来判定。

File: charts/crater/values.yaml [hunk 9]
```diff
@@ -234,6 +234,10 @@ backendConfig:
   # Must be specified for the server to start
   port: :8088
 
+  # -- Network port serving volcano's scheduler extender callbacks (Optional)
+  # Unset deploys without the endpoint; when set it must match extender.urlPrefix in the volcano configmap
+  schedulerExtenderPort: :8089
```
Comment (AkashiSensei) [inline charts/crater/values.yaml:239]: 【优化建议】注释写 Unset 则不部署 endpoint，默认值却是 `:8089`。`example-config.yaml` 里这项是注释掉的。另外 extender 的 reservation ledger 在进程内，不能跨副本共享，backend 必须单副本，Chart 目前没有写出这个约束。

File: cli/skills/crater-cli-job/SKILL.md [hunk 10]
```diff
@@ -48,7 +48,6 @@ crater job ls \
   --search experiment \
   --status Running,Pending \
```
Comment (AkashiSensei) [inline cli/skills/crater-cli-job/SKILL.md:49]: 【核心规范】`crater-cli-job` / `crater-cli-read` 的示例仍用 `--status Running,Pending` 表示运行中和等待中的作业，没有覆盖本 PR 新增的 `Inqueue`。更主要的是，这些 CLI 使用 Skill 没有说明各作业状态的含义，用户和 Agent 无法判断作业当前处于哪个阶段。Skill 内容已改，但 `version` 没有提升。

**Review by AkashiSensei — CHANGES_REQUESTED** [review]
这次把 prequeue watcher 换成 Volcano extender，调度和配额的职责拆分是清楚的，Prequeue drain 也考虑到了升级。

PR 描述没有记录 CLI/后端 `APIVersion` 以及双方最低支持版本是否变化及原因。作业响应新增了 `podGroupPhase`、去掉了 `scheduleType`，管理配置路径也从 `/prequeue` 换成了 `/scheduler-extender`。

File: backend/pkg/reconciler/vcjob-reconciler.go (hunk 1 above)
Comment (YiD11) [reply backend/pkg/reconciler/vcjob-reconciler.go]: 已改。合成的 Inqueue 作业状态和 phase != Pending 的前置判断都去掉了，reconciler 现在总会读取 PodGroup。status 存 vcjob 的原始 phase，新增的 pod_group_phase 存 PodGroup 的原始 phase。"已准入、等待节点"由 utils.IsJobInqueue 根据这两个值判断，空 phase 按 Pending 处理。配额账本在内存里的判断和数据库查询用的是同一个条件，测试里加了空 phase 的用例。

File: backend/cmd/gorm-gen/models/migrate.go (hunk 2 above)
Comment (YiD11) [reply backend/cmd/gorm-gen/models/migrate.go:1636]: 已改。在删表迁移 202609011000 之前新增了 202609010950，把未删除、未过期的 queue_quota_enabled 和 normal_job_waiting_tolerance_seconds 写进 system_configs。旧的配额开关和 backfill_enabled 只要有一个为 true，extender 总开关就打开。system_configs 里已有同名 key 时不覆盖。

File: cli/internal/api/job.go (hunk 3 above)
Comment (YiD11) [reply cli/internal/api/job.go:93]: 已改。后端和 CLI 的 APIVersion 都升到 2，MinSupportedCLIAPIVersion 和 MinSupportedBackendAPIVersion 也都升到 2，原因写在两处常量的注释和 PR 描述里。旧 CLI 会把缺失的 scheduleType 解成 0，这个问题靠最低支持版本处理：v1 CLI 执行 crater compatibility 时会提示 CLI 版本过旧。

File: cli/docs/COMMANDS.md (hunk 4 above)
Comment (YiD11) [reply cli/docs/COMMANDS.md:482]: 已改。删掉了 TensorFlow / PyTorch 不允许 scheduleType=0 的说明和 crater context prequeue，crater admin system-config 的子命令改为 llm|gpu-analysis|scheduler-extender。

File: backend/pkg/extender/server.go (hunk 8 above)
Comment (YiD11) [reply backend/pkg/extender/server.go:25]: 已改。快照不再固定 2 秒失效。只要最近 10 秒内收到过 onSessionClose，快照就保留到下一次 onSessionClose，所以单轮超过 2 秒也不会中途重建。只有连续 10 秒收不到 onSessionClose 时才改为每 2 秒重建一次，并打日志提示，这种情况通常是 volcano configmap 里没配 onSessionCloseVerb。保留这个兜底，是因为 onSessionClose 是可选配置，而且 volcano 发出后不会重试，没有兜底的话快照可能一直不更新。单轮超过 10 秒时仍会退回定时重建。

File: charts/crater/values.yaml (hunk 9 above)
Comment (YiD11) [reply charts/crater/values.yaml:260]: 注释已改，默认值仍是 :8089。默认只创建一个集群内的 ClusterIP Service，在 volcano configmap 的 extender.urlPrefix 指向它、并且管理员打开开关之前，不会有请求打进来。所以这个默认值不改变现有行为，以后启用也不用重新部署 Chart。values.yaml 的注释和 Chart README 已按这个意思重写，并说明了设为 "" 就不部署。example-config.yaml 是本地调试配置的样例，volcano 访问不到本地进程，所以那里保持注释。单副本约束已写进 values.yaml 注释和 Chart README，backend Deployment 的 replicas 在 Chart 里本来就固定为 1。

File: cli/skills/crater-cli-job/SKILL.md (hunk 10 above)
Comment (YiD11) [reply cli/skills/crater-cli-job/SKILL.md:58]: 已改。crater-cli-job 新增了 Job Status 一节，说明每个作业状态的含义，以及作业处于 Pending 时怎样用 podGroupPhase 区分未准入、已准入等待节点和正在启动。crater-cli-read 加了对应的提示。两个 Skill 的 version 分别升到 0.4.0 和 1.3.3。--status Running,Pending 这个示例没改：后续提交已经去掉了 Inqueue 作业状态，已准入、等待节点的作业 status 仍然是 Pending，示例已经覆盖。
